perceiver/data/vision/optical_flow.py

```diff
@@ -0,0 +1,206 @@
+        grid_indices = self._compute_patch_grid_indices(image_pair[0].shape)
+        return self._preprocess(image_pair, grid_indices)
+
+    def preprocess_batch(
+        self, image_pairs: Union[List[Tuple[np.ndarray, np.ndarray]], List[Tuple[torch.Tensor, torch.Tensor]]]
+    ) -> torch.Tensor:
+        shapes = []
+        for image1, image2 in image_pairs:
+            shapes += [image1.shape, image2.shape]
+
+        if not np.all(np.array(shapes) == shapes[0]):
+            raise ValueError("Shapes of images must match. Not all input images have the same shape.")
+
+        grid_indices = self._compute_patch_grid_indices(shapes[0])
+        return torch.stack([self._preprocess(image_pair, grid_indices) for image_pair in image_pairs], dim=0)
+
+    def postprocess(self, predictions: torch.Tensor, img_shape: Tuple[int, ...]) -> torch.Tensor:
+        flow_batch = []
+        height = img_shape[0]
+        width = img_shape[1]
+
+        grid_indices = self._compute_patch_grid_indices(img_shape)
+        prediction_batch = predictions.unsqueeze(0).cpu() if predictions.dim() == 4 else predictions.cpu()
+
+        b, p, *_ = prediction_batch.shape
+        if p != len(grid_indices):
+            raise ValueError(
+                f"Number of patches in the input does not match the number of calculated patches based "
+                f"on the supplied image size (nr_patches='{p}', calculated={len(grid_indices)})."
+            )
+
+        for prediction in prediction_batch:
+            flow = torch.zeros(1, height, width, 2).type(torch.float32)
+            flow_weights = torch.zeros(1, height, width, 1).type(torch.float32)
+
+            for flow_patch, (y, x) in zip(prediction, grid_indices):
+                flow_patch = flow_patch * self.flow_scale_factor
+
+                weights_y, weights_x = torch.meshgrid(
+                    torch.arange(self.patch_size[0]), torch.arange(self.patch_size[1]), indexing="ij"
+                )
+                weights_x = torch.minimum(torch.add(weights_x, 1), self.patch_size[1] - weights_x)
+                weights_y = torch.minimum(torch.add(weights_y, 1), self.patch_size[0] - weights_y)
+                weights = rearrange(torch.minimum(weights_x, weights_y), "h w -> 1 h w 1")
+
+                pad = (0, 0, x, width - x - self.patch_size[1], y, height - y - self.patch_size[0], 0, 0)
+                flow += F.pad(flow_patch * weights, pad, "constant", 0)
+                flow_weights += F.pad(weights, pad, "constant", 0)
+
+            flow /= flow_weights
+            flow_batch.append(flow)
+
+        return torch.concat(flow_batch, dim=0)
+
+    def process(
+        self,
+        model,
+        image_pairs: Union[List[Tuple[np.ndarray, np.ndarray]], List[Tuple[torch.Tensor, torch.Tensor]]],
+        batch_size: int,
+        device: torch.device,
+    ) -> torch.Tensor:
+        image_shape = image_pairs[0][0].shape
+        input_features = self.preprocess_batch(image_pairs)
```

Comment:
This is going to consume a lot of RAM even for `image_pairs` of moderate length and forces a client to provide image_pairs in rather small chunks. What do you think about running `preprocess_batch()` on chunks of `image_pairs` controlled by `batch_size`? For example

```python
    def process(
        self,
        model,
        image_pairs: Union[List[Tuple[np.ndarray, np.ndarray]], List[Tuple[torch.Tensor, torch.Tensor]]],
        batch_size: int,
        device: torch.device,
    ) -> torch.Tensor:
        image_shape = image_pairs[0][0].shape
        grid_indices = self._compute_patch_grid_indices(image_shape)

        predictions = []
        with torch.no_grad():
            for i in tqdm(range(0, len(image_pairs), batch_size)):
                input_features_batch = self._preprocess_batch(image_pairs[i : (i + batch_size)], grid_indices)
                input_features_batch = rearrange(input_features_batch, "b p t c h w -> (b p) t c h w")
                for j in range(0, input_features_batch.shape[0], batch_size):
                    input_features_micro_batch = input_features_batch[j : (j + batch_size)]
                    input_features_micro_batch = input_features_micro_batch.to(device)
                    pred = model(input_features_micro_batch)
                    predictions.append(pred.cpu().detach())
        flow_predictions = torch.concat(predictions, dim=0)
        flow_predictions = rearrange(flow_predictions, "(b p) h w c -> b p h w c", b=len(image_pairs))
        return self.postprocess(flow_predictions, image_shape)
```

This uses a refactored version of `preprocess_batch` to avoid repeated computation of grid indices:

```python
    def preprocess_batch(
        self, image_pairs: Union[List[Tuple[np.ndarray, np.ndarray]], List[Tuple[torch.Tensor, torch.Tensor]]]
    ) -> torch.Tensor:
        grid_indices = self._compute_patch_grid_indices(image_pairs[0][0].shape)
        return self._preprocess_batch(image_pairs, grid_indices)

    def _preprocess_batch(
        self,
        image_pairs: Union[List[Tuple[np.ndarray, np.ndarray]], List[Tuple[torch.Tensor, torch.Tensor]]],
        grid_indices: List[Tuple[int, int]],
    ) -> torch.Tensor:
        shapes = []
        for image1, image2 in image_pairs:
            shapes += [image1.shape, image2.shape]

        if not np.all(np.array(shapes) == shapes[0]):
            raise ValueError("Shapes of images must match. Not all input images have the same shape.")

        return torch.stack([self._preprocess(image_pair, grid_indices) for image_pair in image_pairs], dim=0)
```

These changes would simplify processing the sample video in `inference.ipynb` to 

```python
# sample consecutive video frames
frame_pairs = video_utils.read_video_frame_pairs(Path("sintel_clip_cave_dragon_fight.mp4"))

# process frame_pairs in batches (of size 1 by default) to avoid running out of memory
optical_flows = processor.process(model, frame_pairs, batch_size=1, device=device)

# create video with optical flows
write_optical_flow_video(Path("sintel_clip_cave_dragon_fight_output_new.mp4"), optical_flows)

```

taking the burden from a client cut `frame_pairs` into chunks. RAM usage remains low too.

Reply:
Good idea! We should probably have two different batch sizes though, one controlling the batch size of `image_pairs` (limited by main memory) and one for the model (limited by GPU RAM).

Reply:
I think GPU memory is the limiting factor in almost all cases, hence fine for me to have a single `batch_size` argument.